Add Oracles content (#268)
OffchainLabs/arbitrum-docs: arbitrum-docs/devs-how-tos/how-to-use-oracles.mdx

---
title: "How to use Oracles in Arbitrum"
sidebar_label: Use Oracles in Arbitrum
description: Learn how to integrate Oracles into your dapp
reader-audience: developers who want to build on Ethereum/Arbitrum
content-type: how-to
---

import PublicPreviewBannerPartial from '../partials/_public-preview-banner-partial.md';

<PublicPreviewBannerPartial />

> In this how-to, you’ll learn how to incorporate popular Oracle services into your Arbitrum dApps. Specifically, each section will provide step-by-step instructions on how to obtain current asset prices from oracles, helping you seamlessly integrate these services.

Reviewer: @TucksonDev this is one of my favorite docs PRs ever, for so many reasons:

 - Each of the new docs is clearly a specific type of doc
 - Each doc is clearly aligned with structure conventions for its content-type
 - Clear metadata with a description that will catch search queries
 - We set expectations upfront, with an invitation to provide feedback -> clear CTA
 - Headers are sentence-cased, per conventions

All of that to say, the feedback in this review all falls under the category of "nits" / long-tail optimizations. Love love love the thought put into this reader experience 🙏

Author: Thanks very much! 😃 
I have to say I learned from the best though. Thank _you_ for your constant guidance and support while creating all this content  ❤️

> In this how-to, you’ll learn how to incorporate popular Oracle services into your Arbitrum dApps. Specifically, each section will provide step-by-step instructions on how to obtain current asset prices from oracles, helping you seamlessly integrate these services.

Reviewer: ```suggestion
Oracles are web services that provide a connection between smart contracts and the outside world. They let decentralized apps (dApps) interact with off-chain data and services.

In this how-to, you’ll learn how to fetch the current price of $ARB tokens from a Solidity smart contract using three different oracles: a **Chainlink** oracle, an **API3** oracle, and a **Tellor** oracle.
```

> This guide assumes that you're familiar with the concept of [Oracles](../for-devs/concepts/oracles.mdx) and have a basic understanding of smart contracts and blockchain development. If you’re new to blockchain development, consider going through our [Quickstart: Build a dApp with Arbitrum (Solidity, Hardhat)](../for-devs/quickstart-solidity-hardhat) before proceeding.

Reviewer: ```suggestion
Familiarity with [Oracles](../for-devs/concepts/oracles.mdx), smart contracts, and blockchain development is expected. If you’re new to blockchain development, consider reviewing our [Quickstart: Build a dApp with Arbitrum (Solidity, Hardhat)](../for-devs/quickstart-solidity-hardhat) before proceeding.
```

## Chainlink

> [Chainlink](https://chain.link/) is a widely-recognized Web3 services platform that specializes in decentralized oracle networks. It allows developers to create versatile Web3 applications that can effortlessly connect to real-world data and off-chain computation on Arbitrum and many other blockchain networks. Chainlink offers a variety of data feeds, including asset prices, weather data, random number generation, and many more.

Reviewer: ```suggestion
[Chainlink](https://chain.link/) is a widely-recognized Web3 services platform that specializes in decentralized oracle networks. It lets you build Ethereum and Arbitrum dApps that connect to a variety of off-chain data feeds and APIs, including those that provide asset prices, weather data, random number generation, and more.
```

Definitely feel free to discard/partially incorporate if this edit is too "destructive".

We can now build the function to get the latest price of $ARB. We’ll use this example contract:

```solidity
contract ARBPriceConsumer {
    AggregatorV3Interface internal priceFeed;

    /**
     * Network: Arbitrum One
     * Aggregator: ARB/USD
     * Address: 0xb2A824043730FE05F3DA2efaFa1CBbe83fa548D6
     */
		address constant PROXY = 0xb2A824043730FE05F3DA2efaFa1CBbe83fa548D6;

    constructor() {
        priceFeed = AggregatorV3Interface(PROXY);
    }

    /**
     * Returns the latest price.
     */
    function getLatestPrice() public view returns (int) {
        (
            /* uint80 roundID */,
            int price,
            /*uint startedAt*/,
            /*uint timeStamp*/,
            /*uint80 answeredInRound*/
        ) = priceFeed.latestRoundData();
        return price;
    }
}
```

You can adapt this contract to your needs. Just remember to use the address of the asset you want to request the price for in the appropriate network, and to **deploy your contract to the same network**. Remember we have a [Quickstart](../for-devs/quickstart-solidity-hardhat) available that goes through the process of compiling and deploying a contract.

### More examples

> Refer to [Chainlink’s documentation](https://docs.chain.link/) for more examples for querying price feeds plus other data feeds available.

Reviewer 2: ```suggestion
Refer to [Chainlink’s documentation](https://docs.chain.link/) for more examples of querying price feeds plus other data feeds available.
```

We can now build the function to get the latest price of $ARB. We’ll use this example contract:

```solidity
contract ARBPriceConsumer {
    /**
     * Network: Arbitrum One
     * Aggregator: ARB/USD
     * Proxy: 0x0cB281EC7DFB8497d07196Dc0f86D2eFD21066A5
     */
    address constant PROXY = 0x0cB281EC7DFB8497d07196Dc0f86D2eFD21066A5;

    /**
     * Returns the latest price.
     */
    function getLatestPrice() 
        external
        view
        returns (int224 value, uint256 timestamp)
    {
        (value, timestamp) = IProxy(PROXY).read();
        // If you have any assumptions about `value` and `timestamp`, make sure
        // to validate them right after reading from the proxy.
    }
}
```

You can adapt this contract to your needs. Just remember to use the address of the asset you want to request the price for in the appropriate network, and to **deploy your contract to the same network**. Remember we have a [Quickstart](../for-devs/quickstart-solidity-hardhat) available that goes through the process of compiling and deploying a contract.

### Querying a random number through API3

> [API3 QRNG](https://docs.api3.org/explore/qrng/) is a public utility provided with the courtesy of [Australian National University (ANU)](https://www.anu.edu.au/). It is served as a public good and is free of charge (apart from the gas costs), and it provides quantum randomness when requiring RNG on-chain.

Reviewer 2: ```suggestion
[API3 QRNG](https://docs.api3.org/explore/qrng/) is a public utility provided with the courtesy of [Australian National University (ANU)](https://www.anu.edu.au/). It is served as a public good, it is free of charge (apart from the gas costs), and it provides quantum randomness when requiring RNG on-chain.
```

    function makeRequestUint256() external {
        bytes32 requestId = airnodeRrp.makeFullRequest(
            airnode,
            endpointIdUint256,
            address(this),
            msg.sender,
            address(this),
            this.fulfillUint256.selector,
            ""
        );
        waitingFulfillment[requestId] = true;
        emit RequestedUint256(requestId);
    }

    function fulfillUint256(bytes32 requestId, bytes calldata data)
        external
        onlyAirnodeRrp
    {
        require(
            waitingFulfillment[requestId],
            "Request ID not known"
        );
        waitingFulfillment[requestId] = false;
        uint256 qrngUint256 = abi.decode(data, (uint256));
        
        // Use `qrngUint256` here...
        
        emit ReceivedUint256(requestId, qrngUint256);
    }
}
```

You can adapt this contract to your needs. Just remember to use the addresses of the appropriate network, and to **deploy your contract to the same network**. Remember we have a [Quickstart](../for-devs/quickstart-solidity-hardhat) available that goes through the process of compiling and deploying a contract.

### More examples

> Refer to [API3’s documentation](https://docs.api3.org/) for more examples for querying other data feeds and Airnodes.

Reviewer 2: ```suggestion
Refer to [API3’s documentation](https://docs.api3.org/) for more examples of querying other data feeds and Airnodes.
```
